presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java

```diff
@@ -0,0 +1,481 @@
+                        source);
+
+                // 2. Add aggregation node
+                List<Symbol> groupByKeys = new ArrayList<>();
+                groupByKeys.addAll(groupBySymbols);
+                groupByKeys.add(distinctSymbol);
+                groupByKeys.add(groupSymbol);
+
+                ImmutableMap.Builder aggregationOuputSymbolsMapBuilder = ImmutableMap.builder();
+                AggregationNode aggregationNode = createNonDistinctAggregation(aggregateInfo.get(),
+                        distinctSymbol,
+                        duplicatedDistinctSymbol,
+                        groupByKeys,
+                        groupIdNode,
+                        node,
+                        aggregationOuputSymbolsMapBuilder);
+                // This map has mapping only for aggregation on non-distinct symbols which the new AggregationNode handles
+                Map<Symbol, Symbol> aggregationOuputSymbolsMap = aggregationOuputSymbolsMapBuilder.build();
+
+                 // 3. Add new project node that adds if expressions
+                ProjectNode projectNode = createProjectNode(aggregationNode,
+                        aggregateInfo.get(),
+                        distinctSymbol,
+                        groupSymbol,
+                        aggregationOuputSymbolsMap);
+
+                return projectNode;
+            }
+            return context.defaultRewrite(node, Optional.empty());
+        }
+
+        /*
+         * This Project is useful for cases when we aggregate on distinct and non-distinct values of same symbol, eg:
+         *  select a, sum(b), count(c), sum(distinct c) group by a
+         * Without this Project, we would count additional values for count(c)
+         *
+         * This method also populates maps of old to new symbols. For each key of outputNonDistinctAggregateSymbols,
+         * Higher level aggregation node's aggregaton <key, AggregateExpression> will now have to run AggregateExpression on value of outputNonDistinctAggregateSymbols
+         * Same for outputDistinctAggregateSymbols map
+         */
+        private ProjectNode createProjectNode(AggregationNode source,
+                AggregateInfo aggregateInfo,
+                Symbol distinctSymbol,
+                Symbol groupSymbol,
+                Map<Symbol, Symbol> aggregationOuputSymbolsMap
+                )
+        {
+            ImmutableMap.Builder<Symbol, Expression> outputSymbols = ImmutableMap.builder();
+            ImmutableMap.Builder<Symbol, Symbol> outputNonDistinctAggregateSymbols = ImmutableMap.builder();
+            for (Symbol symbol : source.getOutputSymbols()) {
+                if (distinctSymbol.equals(symbol)) {
+                    Symbol newSymbol = symbolAllocator.newSymbol("expr", symbolAllocator.getTypes().get(symbol));
+                    aggregateInfo.setNewDistinctAggregateSymbol(newSymbol);
+
+                    Expression expression = createIfExpression(groupSymbol.toSymbolReference(),
```

Comment:
This is probably ok for now, but it should eventually be replaced with `GROUPING` when it's available: https://github.com/prestodb/presto/pull/5465